CP-41238: Run commands before collecting directories

bernhardkaindl opened the pull request:
Based on #44 and #50 for working diff-coverage of changed lines.

This PR implements a prototype solution for the problem described below.

The next step would be to write a unit test for GitHub which confirms that the changed order is indeed as intended.

----
When a bug tool is taken, it captures the RRDs for the host. However, as RRDs are only synced to disk periodically, there is a good chance that it doesn't contain anything useful (e.g. if taking a status report immediately after an issue).

It would be useful if the RRDs could be synced to disk during the bug tool. Bugtool also has support for fetching RRDs over HTTP from xAPI,
but this is currently disabled by default.

But as the xen-api added the command `rrd-cli save_rrds` to
```py
/etc/xensource/bugtool/xapi/stuff.xml:<command label="save_rrd">rrd-cli save_rrds</command>
```
The assumption likely was:
- Bugtool will run the command `rrd-cli save_rrds` as defined in the xen-api bugtool xml file (see the code above).
- Bugtool will the read the rrd files.

But the implementation of bugtool is the other way around, to the extreme. Currently, it does this:
 * It looks for which files it can find and records the paths to the files it has found to exist.
 * It reads the files which it found in pass 1 and stores them in memory
 * In pass 3, it executes commands defined by xapi in the xapi bugtool file/plugin.

The bugtool Requirements for the fix may be:
 * Execute rrd-cli save_rrds (or the commands from the plugins such as the xapi plugins at least)
 * This must happen before the directory tree(s) (in this case at least /var/xapi/blobs) are recursively traversed to look for files.
 * This is the pre-condition for rrd-cli to run before the rrd directory ist listed. Only after that the tree /var/xapi/blobs may be collected.

edwintorok commented on xen-bugtool:
```diff
@@ -543,13 +544,42 @@ def file_output(cap, path_list):
+    try:
+        for f in os.listdir(path):
+            fn = os.path.join(path, f)
+            if matches(fn, pattern, negate) and os.path.isfile(fn):
```
Good idea to swap the order here

edwintorok commented on xen-bugtool:
```diff
@@ -572,20 +602,64 @@ def include_inventory(archive, dir):
+    # Then, loop over the files which were found and
+    for k, v in data.items():
+        if "cmd_args" in v:
+            continue  # commands processing has been moved to a differnt loop
```
Typo: different

edwintorok commented on xen-bugtool:
```diff
@@ -572,20 +602,64 @@ def include_inventory(archive, dir):
+    traverse_directory_specifications(directory_specifications, entries)
+
+    # Then, loop over the files which were found and
+    for k, v in data.items():
```
Given this usage the traverse function could be an iterator that calls 'yield' whenever it finds something noteworthy. But its fine this way too.

bernhardkaindl replied:
Yes, this is a great idea! This idea would possibly allow interleaving the directory traversal with the data collection, triggered by yielding each file at the moment it is found, which should lead to possibly more effective (faster) data collection, and file collection and would be interlinked, possibly raising chances of getting new files right when they are getting created.

edwintorok reviewed (commented):
This looks good in general (but I'm not familiar with this codebase), some minor comments.

bernhardkaindl commented on xen-bugtool:
```diff
@@ -543,13 +544,42 @@ def file_output(cap, path_list):
+    :param negate (bool): If True, negate the pattern matching.
+    """
+    try:
+        for f in os.listdir(path):
```
Thanks! That's a good idea for an improvement. I'll note in a PR for planning future work.

edwintorok reviewed (dismissed):
This has been tested in XenRT, and fixed the original bug with the blobs directory not being collected.

There are some CI failures here, I'll kick off a rerun, they seem like some kind of connection errors when submitting coverage reports.

bernhardkaindl commented on xen-bugtool:
```diff
@@ -572,20 +602,64 @@ def include_inventory(archive, dir):
+    # Afterwards, traverse the directory specifications for files to add
+    traverse_directory_specifications(directory_specifications, entries)
+
+    # Then, loop over the files which were found and
```
@psafont Thanks for the review and approve, I added the missing end of the sentence:
```patch
-   # Then, loop over the files which were found and
+   # Then, loop over the files which were found and collect their contents
```
The other changed in the comparision of the new push ar from the merges of the other PRs in the new master onto which I rebased it now too.

Thanks!!!

bernhardkaindl commented:
Thanks for the reviews and approves!
I added the missing end of a sentence in a comment found by @psafont!

And, I applied one part of Edwin's suggestion to not fail CI if the Codecov upload (the one after pre-commit only) fails, because in pre-commit we have a diff-cover job which fails if diff-coverage is missing (tested from triggering it before in this repo).

The other changed in the comparison of the new push are from the merges of the other PRs in the new master onto which I rebased it now too.
